### PR TITLE
1860 Insolvency and train buying rules

### PR DESCRIPTION
**Pull request by roseundy**
Implements all the insolvency rules
Implements train buy restrictions

Common code changes:
- Added ability for show status line for corporations (Game/Base, UI:Corporation)
- Game-specific train owner (Game/Base, Engine/Route)
- Game-specific trains list (Game/Base, UI:Route_Selector)

### CODE REVIEW

File: lib/engine/game/g_1860.rb [hunk 1]
```diff
@@ -175,24 +176,57 @@ def event_fishbourne_to_bank!
...
+      def make_bankrupt(corp)
+        @bankrupt_corps << corp unless bankrupt(corp)
+      end
+
+      def clear_bankrupt(corp)
+        @bankrupt_corps.delete(corp)
+      end
+
+      def status_str(corp)
+        status = nil
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:210]: you don't need to declare variables in ruby like this

File: lib/engine/game/g_1860.rb [hunk 2]
```diff
@@ -269,6 +309,33 @@ def close_other_companies!(company)
...
+        end
+
+        help
+      end
+
+      def train_owner(train)
+        train.owner == @depot ? lesee : train.owner
+      end
+
+      def lesee
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:327]: lesee?

File: lib/engine/game/g_1860.rb [hunk 3]
```diff
@@ -527,23 +613,33 @@ def compute_stops(route)
...
 
         stops
       end
 
       def route_distance(route)
         n_cities = route.stops.select { |n| n.city? || n.offboard? }.size
-        n_towns = route.stops.select { |n| n.town? && !n.halt? }.size
+        # halts are treated like towns for leased trains
+        n_towns = if route.train.owner != @depot
+                    route.stops.select { |n| n.town? && !n.halt? }.size
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:626]: count

File: lib/engine/game/g_1860.rb [hunk 4]
```diff
@@ -527,23 +613,33 @@ def compute_stops(route)
...
+                  else
+                    route.stops.select(&:town?).size
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:628]: count

File: lib/engine/game/g_1860.rb [hunk 5]
```diff
@@ -527,23 +613,33 @@ def compute_stops(route)
...
+          stops.sum { |stop| stop.route_base_revenue(route.phase, route.train) }
+        else
+          40 + 20 * stops.size
+        end
       end
 
-      def subsidy_for(_route, stops)
-        stops.select(&:halt?).size * HALT_SUBSIDY
+      def subsidy_for(route, stops)
+        route.train.owner != @depot ? stops.select(&:halt?).size * HALT_SUBSIDY : 0
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:642]: count

File: lib/engine/game/g_1860.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:210]: OK

File: lib/engine/game/g_1860.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:327]: Should be "lessee" - someone who is a tenant of a lease. Spelling isn't my strong point.

File: lib/engine/game/g_1860.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:626]: OK

File: lib/engine/game/g_1860.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:628]: OK

File: lib/engine/game/g_1860.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:642]: OK. Obviously I have a blind spot about count :-/

**Review by tobymao — APPROVED** [review]

**Pull request merged**